File: Java/1962. Remove Stones to Minimize the Total.java
class Solution {
    public int minStoneSum(int[] piles, int k) {
        PriorityQueue<Integer> q = new PriorityQueue<Integer>(
                Collections.reverseOrder());
        int total = 0, pile;

        for (int p: piles) {
            q.add(p);
            total += p;
        }

        while (k > 0 && !q.isEmpty()) {
            pile = q.poll();
            k--;
            total -= Math.floor(pile / 2);

            if (pile % 2 == 0) {
                pile /= 2;
            } else {
                pile = pile / 2 + 1;
            }

            if (pile > 1) {
                q.add(pile);
            }
        }

        return total;
    }
}

File: Python3/1962. Remove Stones to Minimize the Total.py
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        total = sum(piles)
        piles = [-x for x in piles]
        heapq.heapify(piles)

        while k > 0 and piles:
            pile = -heapq.heappop(piles)
            k -= 1
            total -= pile // 2

            if pile % 2 == 0:
                pile //= 2
            else:
                pile = pile // 2 + 1

            # don't process 1's    
            if pile > 1:
                heapq.heappush(piles, -pile)

        return total